File: dart/tank.dart
import 'dart:html';
import 'dart:math';
import 'package:box2d/box2d_browser.dart';
import 'image.dart';
import 'keyboard.dart';
import 'sprite.dart';
import 'main.dart';

class Tank extends Sprite {
	var accel = 2;
	var maxSpeed = 2, rotateSpeed = 0.75;
	var keyboard;
	var team;
	var shotCooldown = 0.5, lastShot = 0;
	var health = 10;

	Tank(Image img, World world, double startx, double starty) {
		init(img, world, startx, starty);
		keyboard = new Keyboard();

		team = this.img.img.src.contains("green") ? 0 : 1;

		body.fixtureList.userData = "Tank";
	}

	void update() {
		handleKeys();
		updatePosition();
		checkCollision();
	}

	void handleKeys() {
		var forward, backward, right, left, shouldShoot;

		if (team == 0) {
			forward = keyboard.isPressed(KeyCode.UP);
			backward = keyboard.isPressed(KeyCode.DOWN);
			right = keyboard.isPressed(KeyCode.RIGHT);
			left = keyboard.isPressed(KeyCode.LEFT);
			shouldShoot = keyboard.isPressed(KeyCode.SPACE);
		} else {
			forward = keyboard.isPressed(KeyCode.W);
			backward = keyboard.isPressed(KeyCode.S);
			right = keyboard.isPressed(KeyCode.D);
			left = keyboard.isPressed(KeyCode.A);
			shouldShoot = keyboard.isPressed(KeyCode.SHIFT);
		}

		torque = 0.0;

		if (forward) {
			speed -= accel;
		} if (backward) {
			speed += accel;
		}

		if (!forward && !backward) {
			speed = 0;
		}

		if (right) {
			torque = -rotateSpeed;
		} if (left) {
			torque = rotateSpeed;
		}

		if (speed.abs() > maxSpeed) {
			if (speed > 0) {
				speed = maxSpeed;
			} else {
				speed = -maxSpeed;
			}
		}

		if (shouldShoot) {
			shoot();
		}
	}

	void shoot() {
		var curShot = new DateTime.now().millisecondsSinceEpoch;
		if (curShot - lastShot > (shotCooldown * 1000)) {
			lastShot = curShot;
			var x = body.position.x;
			var y = body.position.y;
			addBullet(x + 10.0 * cos(body.angle + PI / 2), y + 10.0 * sin(body.angle + PI / 2), degrees(body.angle));
		}
	}

	void checkCollision() {
		for (ContactEdge ce = body.contactList; ce != null; ce = ce.next) {
			if (ce.contact.touching) {
				var userData = ce.contact.fixtureA.userData;
				var otherUserData = ce.contact.fixtureB.userData;
				if (userData == "Bullet" || otherUserData == "Bullet") {
					health -= 1;
					if (health <= 0) {
						destroy();
					}
				}
				break;
			}
		}
	}
}

File: oldJs/tank.js
Tank = function(img, world, startx, starty, angle, id, mainTank) {
	if (typeof angle === 'undefined') angle = 0;
	this.isMainTank = mainTank;
	this.init(img);
	this.accel = 0.5;
	this.maxSpeed = 2; this.rotateSpeed = 2;
	this.width = 0.75, this.height = 1;
	this.widthScale = 1.4;

	this.id = id;

	this.shotCooldown = 0.5; this.lastShot = 0;
	this.health = 10; this.ammo = 10;
	this.reloadSpeed = 2.5;

	this.createBody(world, startx, starty);
	this.createShape();

	this.body.SetSleepingAllowed(false);
	this.body.SetUserData("Tank");
	this.body.SetLinearDamping(10);
	this.body.SetAngle(angle * toRadians);
}

Tank.prototype = new Sprite();
Tank.prototype.constructor = Tank;

Tank.prototype.update = function(dt) {
	if (this.isMainTank) {
		this.handleKeys();
		this.updatePosition(dt);
	}
	Sprite.prototype.update.call(this, dt);
}

Tank.prototype.handleKeys = function() {
	// Up arrow
	var forward = keysDown[38];
	// Down arrow
	var backward = keysDown[40];
	// Right arrow
	var right = keysDown[39];
	// Left arrow
	var left = keysDown[37];
	// Spacebar
	var shoot = keysDown[32];

	if (forward) {
		this.speed += this.accel;
	} if (backward) {
		this.speed -= this.accel;
	}

	if (this.speed > this.maxSpeed) {
		this.speed = this.maxSpeed;
	} else if (this.speed < -this.maxSpeed) {
		this.speed = -this.maxSpeed;
	}

	if (!(forward || backward)) {
		this.speed = 0;
	}

	if (right) {
		this.torque = this.rotateSpeed;
	} if (left) {
		this.torque = -this.rotateSpeed;
	}

	if (!(right || left)) {
		this.torque = 0;
	}

	if (shoot) {
		this.shoot();
	}

}

Tank.prototype.updatePosition = function(dt) {
	var vx = this.speed * Math.cos(this.body.GetAngle() - Math.PI / 2);
	var vy = this.speed * Math.sin(this.body.GetAngle() - Math.PI / 2);

	var velocity = new b2Vec2(vx * dt, vy * dt);
	this.torque *= dt;

	// this.body.SetLinearVelocity(velocity);
	this.body.ApplyImpulse(velocity, this.body.GetPosition());

	this.body.SetAngularVelocity(this.torque);
}

Tank.prototype.shoot = function() {
	var now = Date.now();
	if (now - this.lastShot > (this.shotCooldown * 1000)) {
		console.log("Pew pew");
		var x = this.body.GetPosition().x;
		var y = this.body.GetPosition().y;
		var angle = this.body.GetAngle();
		var b = {pos: new b2Vec2(x + 2 * Math.cos(angle + Math.PI * 1.5), y + 2 * Math.sin(angle + Math.PI * 1.5)), angle: angle * toDegrees, speed: 0.25};
		addBullet(x + 2 * Math.cos(angle + Math.PI * 1.5), y + 2 * Math.sin(angle + Math.PI * 1.5), angle * toDegrees, 0.25);
		this.lastShot = now;
		socket.emit('shoot', b);
	}
}
